File: matlab/stereo5_BP2.m
% Stereo Matching using Belief Propagation (Synchronous) - smoothness cost function approach
% Computes a disparity map from a rectified stereo pair using Belief Propagation (Synchronous)

function stereo5_BP2()

    % Set parameters
    dispLevels = 16; %disparity range: 0 to dispLevels-1
    lambda = 10; %weight of smoothness cost
    trunc = 2; %truncation of smoothness cost
    iterations = 60;

    % Define matching cost function
    computeMatchingCost = @(left,right) abs(left-right); %absolute differences

    % Define smoothness cost function
    computeSmoothnessCost = @(d1,d2) lambda*min(abs(d1-d2),trunc);

    % Load left and right images in grayscale
    leftImg = rgb2gray(imread('left.png'));
    rightImg = rgb2gray(imread('right.png'));

    % Apply a Gaussian filter
    leftImg = imgaussfilt(leftImg,0.6,'FilterSize',5);
    rightImg = imgaussfilt(rightImg,0.6,'FilterSize',5);

    % Get the size
    [rows,cols] = size(leftImg);

    % Convert to int32
    leftImg = int32(leftImg);
    rightImg = int32(rightImg);

    % Compute pixel-based matching costs (data cost)
    matchingCosts = zeros(rows,cols,dispLevels,'int32');
    for d = 0:dispLevels-1
        rightImgShifted = shiftRight(rightImg,d,0);
        matchingCosts(:,:,d+1) = computeMatchingCost(leftImg,rightImgShifted);
    end

    % Compute smoothness costs
    d = 0:dispLevels-1;
    smoothnessCosts = computeSmoothnessCost(d,d.');
    smoothnessCosts = permute(int32(smoothnessCosts),[3 4 1 2]);

    % Initialize messages for the 4 directions
    fromLeft = zeros(rows,cols,dispLevels,'int32');
    fromRight = zeros(rows,cols,dispLevels,'int32');
    fromUp = zeros(rows,cols,dispLevels,'int32');
    fromDown = zeros(rows,cols,dispLevels,'int32');

    figure
    for it = 1:iterations
        % Create messages to right
        currentCosts = matchingCosts + fromLeft + fromUp + fromDown;
        toRight = computeDirectionalCosts(currentCosts,smoothnessCosts);

        % Create messages to left
        currentCosts = matchingCosts + fromRight + fromUp + fromDown;
        toLeft = computeDirectionalCosts(currentCosts,smoothnessCosts);

        % Create messages to down
        currentCosts = matchingCosts + fromUp + fromLeft + fromRight;
        toDown = computeDirectionalCosts(currentCosts,smoothnessCosts);

        % Create messages to up
        currentCosts = matchingCosts + fromDown + fromLeft + fromRight;
        toUp = computeDirectionalCosts(currentCosts,smoothnessCosts);

        % Send all messages
        fromLeft = shiftRight(toRight,1,0);
        fromRight = shiftLeft(toLeft,1,0);
        fromUp = shiftDown(toDown,1,0);
        fromDown = shiftUp(toUp,1,0);

        % Compute total costs (belief)
        totalCosts = fromLeft + fromRight + fromUp + fromDown;

        % Compute the disparity map
        [~,ind] = min(totalCosts,[],3);
        dispMap = ind-1;

        % Normalize the disparity map for display
        scaleFactor = 256/dispLevels;
        dispImg = uint8(dispMap*scaleFactor);

        % Show disparity map
        imshow(dispImg)

        % Show iterations
        fprintf('iteration: %d/%d\n',it,iterations)
    end

    % Save disparity map
    imwrite(dispImg,'disparity5_BP2.png')
end

% Compute messages
% ----------------
function output = computeDirectionalCosts(currentCosts,smoothnessCosts)
    sum = currentCosts + smoothnessCosts;
    output = permute(min(sum,[],3),[1 2 4 3]);
    output = output - min(output,[],3); %normalize
end

% Shift Functions (Down/Up/Right/Left/Forward/Backward)
% -----------------------------------------------------
function B = shiftDown(A,n,fillValue)
    B = circshift(A,n,1);
    B(1:min(n,end),:,:) = fillValue;
end
function B = shiftUp(A,n,fillValue)
    B = circshift(A,-n,1);
    B(max(end-n+1,1):end,:,:) = fillValue;
end
function B = shiftRight(A,n,fillValue)
    B = circshift(A,n,2);
    B(:,1:min(n,end),:) = fillValue;
end
function B = shiftLeft(A,n,fillValue)
    B = circshift(A,-n,2);
    B(:,max(end-n+1,1):end,:) = fillValue;
end
function B = shiftForward(A,n,fillValue)
    B = circshift(A,n,3);
    B(:,:,1:min(n,end)) = fillValue;
end
function B = shiftBackward(A,n,fillValue)
    B = circshift(A,-n,3);
    B(:,:,max(end-n+1,1):end) = fillValue;
end

File: python/stereo5_BP2.py
# Stereo Matching using Belief Propagation (Synchronous) - smoothness cost function approach
# Computes a disparity map from a rectified stereo pair using Belief Propagation (Synchronous)

import numpy as np
import cv2 as cv
import matplotlib.pyplot as plt

# Set parameters
dispLevels = 16 #disparity range: 0 to dispLevels-1
lambda_ = 10 #weight of smoothness cost
trunc = 2 #truncation of smoothness cost
iterations = 60

# Define matching cost function
computeMatchingCost = lambda left,right: np.absolute(left-right) #absolute differences

# Define smoothness cost function
computeSmoothnessCost = lambda d1,d2: lambda_*np.minimum(np.absolute(d1-d2),trunc)

# Compute messages
def computeMinSumCosts(costs):
    sumCosts = costs[:,:,:,np.newaxis] + smoothnessCosts4d
    minSumCosts = np.amin(sumCosts,axis=2)
    minSumCosts = minSumCosts - np.amin(minSumCosts,axis=2)[:,:,np.newaxis] #normalize messages
    return minSumCosts

# Load left and right images in grayscale
leftImg = cv.imread("left.png",cv.IMREAD_GRAYSCALE)
rightImg = cv.imread("right.png",cv.IMREAD_GRAYSCALE)

# Apply a Gaussian filter
leftImg = cv.GaussianBlur(leftImg,(5,5),0.6)
rightImg = cv.GaussianBlur(rightImg,(5,5),0.6)

# Get the size
(rows,cols) = leftImg.shape

# Convert to int32
leftImg = leftImg.astype(np.int32)
rightImg = rightImg.astype(np.int32)

# Compute pixel-based matching costs (data cost)
matchingCosts = np.zeros((rows,cols,dispLevels),dtype=np.int32)
for d in range(dispLevels):
    rightImgShifted = np.roll(rightImg,d,1)
    matchingCosts[:,:,d] = computeMatchingCost(leftImg,rightImgShifted)

# Compute smoothness costs
d = np.arange(dispLevels)
smoothnessCosts = computeSmoothnessCost(d,d[np.newaxis,:].T)
smoothnessCosts4d = smoothnessCosts[np.newaxis,np.newaxis,:,:].astype(np.int32)

# Initialize messages for the 4 directions
fromLeft = np.zeros((rows,cols,dispLevels),dtype=np.int32)
fromRight = np.zeros((rows,cols,dispLevels),dtype=np.int32)
fromUp = np.zeros((rows,cols,dispLevels),dtype=np.int32)
fromDown = np.zeros((rows,cols,dispLevels),dtype=np.int32)

for it in range(iterations):
    # Create messages to right
    costs = matchingCosts + fromLeft + fromUp + fromDown
    toRight = computeMinSumCosts(costs)

    # Create messages to left
    costs = matchingCosts + fromRight + fromUp + fromDown
    toLeft = computeMinSumCosts(costs)

    # Create messages to down
    costs = matchingCosts + fromUp + fromLeft + fromRight
    toDown = computeMinSumCosts(costs)

    # Create messages to up
    costs = matchingCosts + fromDown + fromLeft + fromRight
    toUp = computeMinSumCosts(costs)

    # Send all messages
    fromLeft = np.roll(toRight,1,1) #shift right
    fromRight = np.roll(toLeft,-1,1) #shift left
    fromUp = np.roll(toDown,1,0) #shift down
    fromDown = np.roll(toUp,-1,0) #shift up

    # Compute total costs (belief)
    totalCosts = fromLeft + fromRight + fromUp + fromDown

    # Compute the disparity map
    dispMap = np.argmin(totalCosts,axis=2)

    # Normalize the disparity map for display
    scaleFactor = 256/dispLevels
    dispImg = (dispMap*scaleFactor).astype(np.uint8)

    # Show disparity map
    plt.cla()
    plt.imshow(dispImg,cmap="gray")
    plt.show(block=False)
    plt.pause(0.01)

    # Show iterations
    print("iteration: {0}/{1}".format(it+1,iterations))

# Save disparity map
cv.imwrite("disparity5_BP2.png",dispImg)

plt.show()
